### PR TITLE
AUTH-26 CSR approving mechanism

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.redhat.com/browse/AUTH-26

This Pull Request introduces CSR Approver (see https://github.com/openshift/library-go/pull/1083) into the Cluster Policy Controller.

### CODE REVIEW

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [inline pkg/cmd/controller/csr.go:38]: @stlaz @s-urbaniak Both the CSR Filter and the Service Account Approver seem to do the same thing (in terms of functionality) - they filter somehow what CSRs are being approved. Perhaps we should create a Service Account Filter for them and squash both things into a single interface?

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [inline pkg/cmd/controller/csr.go:18]: @s-urbaniak Please check those 3 items - are they correct?

File: pkg/cmd/controller/csr.go
Comment (stlaz) [reply pkg/cmd/controller/csr.go:38]: They each serve a different purpose, I would be against such a unification

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:37]: I hope this is a subject to change :slightly_smiling_face:

File: pkg/cmd/controller/csr.go
Comment (stlaz) [reply pkg/cmd/controller/csr.go:18]: the subject seems wrong, I would expect it to match the name of the serviceaccount pulling the metrics

Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:18]: Yes, the certificate requestor is cluster-monitoring-operator, the target subject is Prometheus, how about `/CN=prometheus`?

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:37]: what do we want to set here? :thinking:

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:38]: agreeing with @stlaz, a service account approver is just a very specific filter validating username, groups, and subject.

Comment (slaskawi) [reply pkg/cmd/controller/csr.go:38]: Ok, understood. Thanks for the context!

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:37]: Exactly! So what shall we put here? @stlaz?

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:18]: Done

**Comment by slaskawi** [comment]
Rebased and updated. Ready for the first round of reviews.

File: pkg/cmd/controller/csr.go
Comment (stlaz) [reply pkg/cmd/controller/csr.go:37]: configure a specific name/label for the CSR

**Comment by soltysh** [comment]
@atiratree fyi, since you were to bump the deps.

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:37]: Oh, so an opinioated one for the sake of this use case? of course, that makes sense :+1: 
This would prevent auto-approving other CSRs from the same serviceaccont.
How about `openshift.io/metrics`?

**Comment by atiratree** [comment]
@soltysh we still want to update k8s to v0.22 once it is released, right?

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go]: Ok, updated. Please take another look at it.

**Comment by slaskawi** [comment]
@s-urbaniak @stlaz Ready for the next round of reviews.

File: pkg/operator/operatorclient.go
Comment (s-urbaniak) [inline pkg/operator/operatorclient.go:10]: nit: whitespace in imports (really a very small nit)

**Comment by stlaz** [comment]
/hold
there shall be no jira ticket names in commit messages

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:10]: nit: order this such that it's
```
system libs

k8s libs

openshift libs

local libs
```

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:19]: @s-urbaniak I would've expected `/CN=system:serviceaccount:openshift-monitoring:openshift-state-metrics`, no?

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:28]: we'll want to get https://github.com/openshift/library-go/pull/1105 merged first to use it here

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:33]: how about `metrics.openshift.io/csr.subject: openshift-state-metrics`?

File: pkg/operator/operatorclient.go
Comment (stlaz) [inline pkg/operator/operatorclient.go:5]: nit: split system and k8s deps with a newline

File: pkg/operator/operatorclient.go
Comment (stlaz) [inline pkg/operator/operatorclient.go:36]: we don't need this if we get https://github.com/openshift/library-go/pull/1105 merged first

**Review by stlaz — CHANGES_REQUESTED** [review]
we'll want to find an approver for https://github.com/openshift/library-go/pull/1105 to include it here

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:19]: sounds good to me. **does** it have to match the exact service account name? I may be missing something, I assumed this is more or less free-form.

Comment (stlaz) [reply pkg/cmd/controller/csr.go:19]: I don't think it has to match, although I think I would expect it to. Or at least prepend it with the `system:` prefix

Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go:19]: I think following the canonical k8s syntax as you suggested sounds good. It is very clear to what it maps to :+1:

**Comment by slaskawi** [comment]
> we'll want to find an approver for openshift/library-go#1105 to include it here

@stlaz I see there are some outstanding comments there, so let me propose to implement this in two-steps:
- We'll leave the code as is
- Once https://github.com/openshift/library-go/pull/1105 gets merged, I'll update the code here.

Would this work for you?

File: pkg/operator/operatorclient.go
Comment (slaskawi) [reply pkg/operator/operatorclient.go:5]: I see @s-urbaniak has commented the other file because there was a blank line :)

@stlaz @s-urbaniak @sttts Do we have a formatter for our IDEs somewhere or at least written guidelines about the code style? Sorting imports manually is a wasteful job. 

If there is no formatter and no written code style guidelines, I suggest to leave this as is and introduce the `goimports` into our Makefiles (of course I volunteer to do that).

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:10]: Let's continue this discussion here: https://github.com/openshift/cluster-policy-controller/pull/61#discussion_r654274660

File: pkg/operator/operatorclient.go
Comment (slaskawi) [reply pkg/operator/operatorclient.go:10]: Let's continue this discussion here: https://github.com/openshift/cluster-policy-controller/pull/61#discussion_r654274660

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:19]: Done

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:28]: I proposed slightly different (more async) approach to this here: https://github.com/openshift/cluster-policy-controller/pull/61#issuecomment-863886456

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:33]: Done

File: pkg/operator/operatorclient.go
Comment (slaskawi) [reply pkg/operator/operatorclient.go:36]: I proposed slightly different (more async) approach to this here: https://github.com/openshift/cluster-policy-controller/pull/61#issuecomment-863886456

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:11]: wrong order

File: pkg/operator/operatorclient.go
Comment (stlaz) [reply pkg/operator/operatorclient.go:36]: unfortunately that does not help, as long as the errors are returned here we're not moving forward by merging this PR

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/operator/operatorclient.go
Comment (slaskawi) [reply pkg/operator/operatorclient.go:5]: It turned out that @stlaz uses `goimports`. I'll give it a go.

File: pkg/operator/operatorclient.go
Comment (slaskawi) [reply pkg/operator/operatorclient.go:36]: @stlaz @s-urbaniak So it seems this PR is blocked until https://github.com/openshift/library-go/pull/1105 gets in.

Comment (slaskawi) [reply pkg/operator/operatorclient.go:36]: Just in case - I copy-pasted the `NoOpOperatorClient` temporarily into my PR as well.

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:11]: Fixed

**Comment by soltysh** [comment]
> @soltysh we still want to update k8s to v0.22 once it is released, right?

Correct, but it should be easier after this :wink:

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [inline pkg/cmd/controller/csr.go:20]: as discussed in https://github.com/openshift/cluster-policy-controller/pull/61#discussion_r649790728 this should rather be `/CN=system:serviceaccount:openshift-monitoring:prometheus`

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [inline pkg/cmd/controller/csr.go:17]: reference the commit id as master will change https://github.com/openshift/cluster-monitoring-operator/blob/e7a2b2e20d19ba7a9271df59e06b1fd32b8bdfe4/manifests/0000_50_cluster-monitoring-operator_03-role-binding.yaml#L25-L27

File: pkg/cmd/controller/csr.go
Comment (s-urbaniak) [inline pkg/cmd/controller/csr.go:22]: `prometheus`

Comment (slaskawi) [reply pkg/cmd/controller/csr.go:22]: Done

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:17]: No, I'm not very keen on doing this. If this file gets moved, the ID as SHA-1 will point to a still-existing file. This will give proper pointers where to look next.

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:20]: Done

**Comment by slaskawi** [comment]
@stlaz @s-urbaniak Updated and rebased. Ready for the next round.

File: pkg/cmd/controller/csr.go
Comment (stlaz) [reply pkg/cmd/controller/csr.go]: actually, according to https://github.com/openshift/library-go/blob/master/pkg/authorization/hardcodedauthorizer/metrics.go#L23, we may want to just go with `"system:serviceaccount:openshift-monitoring:prometheus-k8s"`

Comment (s-urbaniak) [reply pkg/cmd/controller/csr.go]: ah yes, i missed the suffix

Comment (slaskawi) [reply pkg/cmd/controller/csr.go]: Done

**Comment by slaskawi** [comment]
Updated and rebased

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:33]: nit: `s/selection.In/selection.Equals/`

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:37]: I believe this should be `selector = selector.Add(*labelsRequirement)`

https://github.com/enj/cluster-authentication-operator/blob/d303865704cb8d9e0e807e63d2f578ecebb42aa2/vendor/k8s.io/apimachinery/pkg/labels/selector.go#L365-L376

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:37]: :man_facepalming: Of course! You're right!

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:33]: Done

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:37]: Done

File: pkg/cmd/controller/csr.go
Comment (stlaz) [inline pkg/cmd/controller/csr.go:16]: nittiest of nits: let's not put github references in the code, I think the hardcode is apparent here.

btw: I believe the correct reference would be https://github.com/openshift/cluster-monitoring-operator/blob/e7a2b2e20d19ba7a9271df59e06b1fd32b8bdfe4/manifests/0000_50_cluster-monitoring-operator_03-role-binding.yaml#L1-L10

**Comment by stlaz** [comment]
another nit: make the commit message be `CSR approver for monitoring scraper`

File: pkg/cmd/controller/csr.go
Comment (slaskawi) [reply pkg/cmd/controller/csr.go:16]: Removed that comment.

**Review by soltysh — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
